granary/reddit.py

```diff
@@ -0,0 +1,218 @@
+
+    Args:
+      user: praw Redditor object
+
+    Returns:
+      an ActivityStreams actor dict, ready to be JSON-encoded
+    """
+    username = user.name
+    if not username:
+      return {}
+
+
+    # trying my best to grab all the urls from the profile description
+
+    description = ''
+    if user.subreddit:
+      user_url = self.BASE_URL + user.subreddit.get('url')
+      urls = [user_url]
+      description = user.subreddit['public_description']
+      profile_urls = util.extract_links(description)
+      urls += util.trim_nulls(profile_urls)
+    else:
+      urls = [self.BASE_URL + '/user/' + username]
+
+    image = user.icon_img
+
+    return util.trim_nulls({
+      'objectType': 'person',
+      'displayName': username,
+      'image': {'url': image},
+      'id': self.tag_uri(username),
+      # numeric_id is our own custom field that always has the source's numeric
+      # user id, if available.
+      'numeric_id': user.id,
+      'published': util.maybe_timestamp_to_iso8601(user.created_utc),
+      'url': urls[0],
+      'urls': [{'value': u} for u in urls] if len(urls) > 1 else None,
+      'username': username,
+      'description': description,
+      })
+
+
+  def praw_to_object(self, thing, type):
+    """
+    Converts a praw object to an object.
+
+    Args:
+      thng: a praw object, Sumbission or Comment
+      type: string to denote whether to get submission or comment content
+
+    Returns:
+      an ActivityStreams object dict, ready to be JSON-encoded
+      """
+    obj = {}
+
+    # always prefer id_str over id to avoid any chance of integer overflow.
+    # usually shouldn't matter in Python, but still.
+    id = thing.id
+    if not id:
+      return {}
+
+    published = util.maybe_timestamp_to_iso8601(thing.created_utc)
+
+    obj = {
+      'id': self.tag_uri(id),
+      'objectType': 'note',
+      'published': published,
+      'to': [],
+      }
+
+    user = thing.author
+    if user:
+      obj['author'] = self.user_to_actor(user)
+      username = obj['author'].get('username')
+
+    obj['to'].append({
+      'objectType': 'group',
+      'alias': '@public',
+      })
+
+    obj['url'] = self.BASE_URL + thing.permalink
+
+    if type == 'submission':
+      obj['content'] = thing.title
+    elif type == 'comment':
+      obj['content'] = thing.body
+
+    return self.postprocess_object(obj)
+
+
+  def praw_to_activity(self, thing, type):
+    """Converts a praw submission or comment to an activity.
+    https://praw.readthedocs.io/en/latest/code_overview/models/submission.html
+    https://praw.readthedocs.io/en/latest/code_overview/models/comment.html
+
+    Args:
+      thng: a praw object, Sumbission or Comment
+      type: string to denote whether to get submission or comment content
+
+    Returns:
+      an ActivityStreams activity dict, ready to be JSON-encoded
+    """
+    obj = self.praw_to_object(thing, type)
+    actor = obj['author']
```

Comment:
in general, you'll want to mostly use `.get()` instead of `[]` in most places in granary, except if you can guarantee the field is always there. (and even then, often prefer `.get()` because someone might later change the code that populates it, and they won't know to update your consuming 
code.)

eg you need it here because `praw_to_object` only populates `author` if the user is available.